acme/acme/standalone.py

```diff
@@ -120,6 +127,45 @@ def shutdown_and_server_close(self):
         self.threads = []
 
 
+class BadALPNProtos(Exception):
+    """Error raised when cannot negotiate ALPN protocol."""
+
+
+class TLSALPN01Server(TLSServer, ACMEServerMixin):
+    """TLSALPN01 Server."""
+
+    ACME_TLS_1_PROTOCOL = b"acme-tls/1"
+
+    def __init__(self, server_address, certs, challenge_certs, ipv6=False):
+        TLSServer.__init__(
+            self, server_address, _BaseRequestHandlerWithLogging, certs=certs,
+            ipv6=ipv6)
+        self.challenge_certs = challenge_certs
+
+    def _cert_selection(self, connection):
+        # TODO: We would like to serve challenge cert only if asked for it via
+        # ALPN. To do this, we need to retrieve the list of protos from client
+        # hello, but this is currently impossible with openssl [0], and ALPN
+        # negotiation is done after cert selection.
+        # Therefore, currently we always return challenge cert, and terminate
+        # handshake in alpn_selection() if ALPN protos are not what we expect.
+        # [0] https://github.com/openssl/openssl/issues/4952
+        server_name = connection.get_servername()
+        logger.debug("Serving challenge cert for server name %s", server_name)
+        return self.challenge_certs.get(server_name, None)
+
+    def _alpn_selection(self, connection, alpn_protos):
+        """Callback to select alpn protocol."""
+        if len(alpn_protos) == 1 and alpn_protos[0] == self.ACME_TLS_1_PROTOCOL:
+            logger.debug("Agreed on %s ALPN", self.ACME_TLS_1_PROTOCOL)
+            return self.ACME_TLS_1_PROTOCOL
+        logger.debug("Cannot agree on ALPN proto. Got: %s", str(alpn_protos))
+        # Explicitly close the connection now because some old versions of OpenSSL do
```

Comment:
According to https://www.pyopenssl.org/en/stable/api/ssl.html#OpenSSL.SSL.Context.set_alpn_select_callback, if we want to close the connection we should return an empty bytestring here. I tested this on a few systems and it seemed to always work.

Unless you can find a problem with this approach, I think we should replace the rest of this function with `return b""` and comment explaining what this return value does. If we take this approach, the exception `BadALPNProtos` will no longer be used so I think we should delete it.

Reply:
Yes you are right. The use of an exception was in the initial PR I took over, but it seems that it was working because recent versions of pyopenssl was properly handling exceptions by returning the empty string, while old ones were not.

Anyway the spec clearly indicates to return an empty string, so better to follow that.